031-statefulset-removal.md

```diff
@@ -0,0 +1,242 @@
+But these are dependent on the current design of the StatefulSet controller.
+Kubernetes changes have caused issues to this feature in the past.
+For example, to add volumes to JBOD storage, we currently need to use a special rolling update which does not roll the controller Kafka broker last.
+
+## Motivation
+
+Some of the limitations described above are blocking us in further development of features that would be hard to implement with StatefulSets.
+Therefore it was decided to investigate what options we have to replace StatefulSets to gain more flexibility and be able to continue the Strimzi development.
+Below are some of the features and use cases which might be significantly easier to implement without the StatefulSet limitations.
+
+*Note: This proposal does not cover these features and use cases.
+They are just used to explain the things which might be easier when this proposal is implemented.*
+
+### Individual broker configuration
+
+Being able to individually configure the brokers would allow us to improve Strimzi in several areas:
+* Less rolling updates
+* Asymmetric configurations of storage or CPU / memory resources
+
+### ZooKeeper-less Kafka
+
+The removal of ZooKeeper from Kafka will bring a lot of advantages but also some challenges.
+There won't be a clear separation between the ZooKeeper cluster and the Kafka cluster.
+Any Kafka broker can be in theory configured as controller.
+With a single StatefulSet managing all Kafka pods, it would be hard to handle all the possible configurations:
+* Dedicated set of controller nodes
+* All nodes being both controllers and brokers
+* Some nodes being both brokers and controller and some only brokers
+
+We could use multiple StatefulSets to separate the different types of nodes.
+But even then it would be hard to change the cluster topologies and the node types.
+Without the StatefulSet limitations, we should have a lot more flexibility to mix the node types and their roles.
+
+### Stretch clusters
+
+Pods managed by a single StatefulSet always exist only on single Kubernetes cluster.
+If you want to have a Kafka cluster which would stretch multiple Kubernetes clusters, you would need to create a StatefulSet on each of them and have it create the Pods on each given cluster.
+But you would not be able easily move the brokers between the Kubernetes clusters since they would depend on the StatefulSet index numbers to generate their broker IDs.
+By introducing _StrimziPodSets_ (for more details and explanation, check the _Proposal_ section below), you can have different StrimziPodSets in each Kubernetes cluster and just delete a pod on one Kubernetes cluster and move it to another cluster by modifying the corresponding StrimziPodSets.
+
+## Proposal
+
+The StatefulSets should be replaced with a new Strimzi custom resource called `StrimziPodSet`.
+StrimziPodSet will be used only internally.
```

Comment:
When we say this, are we expecting to have the same compatibility guarantees for `StrimziPodSet` as for the other resources (I think we are, but I just want everyone to be clear about this).

Reply:
Well, I think anything can happen during the alpha phase. Once it leaves the alpha phase, it should be treated as any other resource with backwards compatibility, versioning etc. I added such explanation to the text as well.